File: ocaml/bench/diff_cadeque.ml
(* diff_cadeque.ml — deterministic CATENABLE workload + per-slot sequences.
 *
 * The companion C driver (c/tests/diff_cadeque.c) generates the SAME
 * workload (same xorshift64 seed/sequence, same op decoding) against the
 * hand-written C catenable port, and emits the same text.  `diff` the two
 * outputs to detect any divergence between the C port and this verified
 * Coq-extracted artifact (KTFlatCadeque, keystone in
 * rocq/KTDeque/Catenable/FlatKeystone.v).
 *
 * Must stay byte-for-byte in step with the C driver: same PRNG, same
 * number of PRNG draws per step, same op decoding, same output format. *)

module C = KTFlatCadeque

let xs = ref 0L
let xnext () =
  let open Int64 in
  let x = !xs in
  let x = logxor x (shift_left x 13) in
  let x = logxor x (shift_right_logical x 7) in
  let x = logxor x (shift_left x 17) in
  xs := x; x

(* unsigned mod K of a 64-bit value *)
let umod (x : int64) (k : int) : int =
  Int64.to_int (Int64.unsigned_rem x (Int64.of_int k))

let push x d = C.cad_push_x x d
let inject d x = C.cad_inject_x d x
let pop d = match C.cad_pop_x d with Some (x, d') -> Some (x, d') | None -> None
let eject d = match C.cad_eject_x d with Some (d', x) -> Some (d', x) | None -> None
let concat a b = match C.cad_concat_x a b with
  | Some c -> c
  | None -> failwith "cad_concat_x: None on regular inputs"

let () =
  let n = if Array.length Sys.argv > 1 then int_of_string Sys.argv.(1) else 100000 in
  let seed =
    if Array.length Sys.argv > 2 then
      Int64.of_string ("0x" ^ Sys.argv.(2))
    else 0x123456789abcdef0L in
  let k = if Array.length Sys.argv > 3 then int_of_string Sys.argv.(3) else 8 in
  let k = if k < 2 then 2 else k in
  xs := seed;

  let slot = Array.make k C.fcad_empty in
  let ctr = ref 1 in
  for _ = 1 to n do
    let op = umod (xnext ()) 5 in
    let i = umod (xnext ()) k in
    (match op with
     | 0 -> slot.(i) <- push !ctr slot.(i); incr ctr
     | 1 -> slot.(i) <- inject slot.(i) !ctr; incr ctr
     | 2 -> (match pop slot.(i) with Some (_, d') -> slot.(i) <- d' | None -> ())
     | 3 -> (match eject slot.(i) with Some (d', _) -> slot.(i) <- d' | None -> ())
     | _ ->
        (* j <> i: self-concat would double the length (exponential) *)
        let j = (i + 1 + umod (xnext ()) (k - 1)) mod k in
        slot.(i) <- concat slot.(i) slot.(j);
        slot.(j) <- C.fcad_empty)
  done;

  let buf = Buffer.create (1 lsl 20) in
  let total = ref 0 in
  for i = 0 to k - 1 do
    (* length via a pop-drain count on a copy *)
    let rec len acc d = match pop d with Some (_, d') -> len (acc+1) d' | None -> acc in
    Buffer.add_string buf (Printf.sprintf "SLOT %d [%d]:" i (len 0 slot.(i)));
    let rec drain d = match pop d with
      | Some (x, d') -> Buffer.add_string buf (Printf.sprintf " %d" x); incr total; drain d'
      | None -> () in
    drain slot.(i);
    Buffer.add_char buf '\n'
  done;
  Buffer.add_string buf (Printf.sprintf "DONE total=%d\n" !total);
  print_string (Buffer.contents buf)

File: c/tests/diff_cadeque.c
/* diff_cadeque.c — deterministic CATENABLE workload + per-slot sequences.
 *
 * The companion OCaml driver (ocaml/bench/diff_cadeque.ml) generates the
 * SAME workload (same xorshift64 seed/sequence, same op decoding) against
 * the Coq-extracted KTFlatCadeque, and emits the same text.  `diff` the
 * two outputs to detect any divergence between the C catenable port and
 * the verified OCaml artifact.
 *
 * Workload: K register slots, each a persistent cadeque.  Each step the
 * PRNG picks an op and slot(s):
 *   0 push   ctr -> slot i        1 inject slot i <- ctr
 *   2 pop    slot i               3 eject  slot i
 *   4 concat slot i := i ++ j ;  if i!=j then slot j := empty
 * Values are a monotone counter so order is checkable.  Both drivers call
 * the PRNG the same number of times in the same order.
 *
 * Output: one line per slot, "SLOT i [n]: v1,v2,...", the slot drained
 * front-to-back via kc_pop (so the emission itself exercises pop), then
 * "DONE total=T".
 *
 * Args: ./diff_cadeque [N] [SEED_HEX] [K]
 */
#include "ktcadeque.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

static uint64_t xs;
static uint64_t xnext(void) {
    uint64_t x = xs;
    x ^= x << 13; x ^= x >> 7; x ^= x << 17;
    xs = x; return x;
}

int main(int argc, char** argv) {
    long N = argc > 1 ? atol(argv[1]) : 100000;
    xs     = argc > 2 ? strtoull(argv[2], NULL, 16) : 0x123456789abcdef0ULL;
    int K  = argc > 3 ? atoi(argv[3]) : 8;
    if (K < 2) K = 2;

    kc_cadeque* slot = malloc((size_t)K * sizeof(kc_cadeque));
    for (int i = 0; i < K; i++) slot[i] = kc_empty();

    long ctr = 1;
    for (long step = 0; step < N; step++) {
        int op = (int)(xnext() % 5);
        int i  = (int)(xnext() % (uint64_t)K);
        switch (op) {
            case 0: slot[i] = kc_push((kt_elem)(intptr_t)(ctr++), slot[i]); break;
            case 1: slot[i] = kc_inject(slot[i], (kt_elem)(intptr_t)(ctr++)); break;
            case 2: { kt_elem e; int ok; slot[i] = kc_pop(slot[i], &e, &ok); break; }
            case 3: { kt_elem e; int ok; slot[i] = kc_eject(slot[i], &e, &ok); break; }
            case 4: {
#ifdef KC_COMPACT_EVERY
                if ((step % KC_COMPACT_EVERY) == 0) kc_arena_compact(slot, K);
#endif
                /* j != i: self-concat would double the sequence length
                 * (exponential blow-up over many steps) — exclude it so the
                 * workload stays linear-sized.  One PRNG draw, K-1 choices. */
                int j = (i + 1 + (int)(xnext() % (uint64_t)(K - 1))) % K;
                slot[i] = kc_concat(slot[i], slot[j]);
                slot[j] = kc_empty();
                break;
            }
        }
    }

    long total = 0;
    for (int i = 0; i < K; i++) {
        size_t n = kc_length(slot[i]);
        printf("SLOT %d [%zu]:", i, n);
        kc_cadeque d = slot[i];
        for (;;) {
            kt_elem e; int ok;
            d = kc_pop(d, &e, &ok);
            if (!ok) break;
            printf(" %ld", (long)(intptr_t)e);
            total++;
        }
        printf("\n");
    }
    printf("DONE total=%ld\n", total);
    free(slot);
    return 0;
}
